homeassistant/helpers/template.py

```diff
@@ -141,6 +145,34 @@ def async_render(self, variables: TemplateVarsType = None,
         except jinja2.TemplateError as err:
             raise TemplateError(err)
 
+    def render_with_collect(
+            self, variables: TemplateVarsType = None,
+            **kwargs) -> (Union[str, TemplateError],
+                          EntityFilter):
+        """Render given template and collect an entity filter."""
+        if variables is not None:
+            kwargs.update(variables)
+
+        return run_callback_threadsafe(
+            self.hass.loop, self.async_render_with_collect,
+            kwargs).result()
+
+    @callback
+    def async_render_with_collect(
+            self, variables: TemplateVarsType = None,
+            **kwargs) -> (Union[str, TemplateError],
+                          EntityFilter):
+        """Render the template and collect an entity filter."""
+        assert self.hass and _COLLECT_KEY not in self.hass.data
+        entity_collect = self.hass.data[_COLLECT_KEY] = EntityFilter()
```

Comment:
Why do we create a new one? Shouldn't we want to collect into the old filter ? If we're not changing filters, then why are we rewriting EntityFilter to be mutable to begin with?

Reply:
The filter is associated with the output of a particular run of the template. If states change, the filter can change. The algorithm is explained in the arch issue.

Reply:
This is just storing the filter in data for this particular run only. It doesn't stay there indefinitely. It's safe to do this because async methods are all executed atomically and single threaded with respect to a particular hass object. It's like passing it through as an extra method parameter deep into the call stack.